genesis/model/contract/smart_contract.go

```diff
@@ -37,16 +39,31 @@ type BaseSmartContract struct {
 	ChildStorage   storage.Storage
 	ContextStorage storage.Storage
 	Parent         object.Parent
+	Resolver       resolver.Resolver
 }
 
 // NewBaseSmartContract creates new BaseSmartContract instance with empty CompositeMap, ChildStorage and specific parent.
 func NewBaseSmartContract(parent object.Parent) *BaseSmartContract {
 	// TODO: NewCompositeHolder
-	return &BaseSmartContract{
+	sc := BaseSmartContract{
 		CompositeMap: make(map[string]object.Composite),
 		ChildStorage: storage.NewMapStorage(),
 		Parent:       parent,
 	}
+	sc.InitiateResolver(sc)
+	return &sc
+}
+
+// InitiateResolver create resolver.
+func (sc *BaseSmartContract) InitiateResolver(self interface{}) {
```

Comment:
You can also rename it to GetResolver() and make resolver field private. GetResolver method will act like singleton - return instance or create it if necessary